### DISCUSSION TITLE
Request for Access to Source Code or MOOSE/MARMOT Permissions for CP-PF coupled model

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#25687] [Q&A General]
Dear MOOSE Community,

I am a researcher interested in the topic of grain boundary migration and its connection to plasticity deformation. I have recently come across the paper titled "[On the plastic driving force of grain boundary migration: A fully coupled phase field and crystal plasticity model](https://www.sciencedirect.com/science/article/pii/S092702561630605X#s0040)," which has sparked my interest in further developing a grain boundary migration model that incorporates fatigue loading.

I am reaching out to inquire if it would be possible to access the relevant source code or obtain the permission for using MOOSE/MARMOT, as mentioned in the aforementioned paper. The insights and techniques discussed in the paper have provided valuable inspiration for my research, and I believe that access to the code or MOOSE/MARMOT would significantly facilitate model development in this part.

I understand the importance of respecting intellectual property and licensing agreements, and I am fully committed to adhering to any terms and conditions associated with the use of the code or software. My research aims to contribute to the understanding of grain boundary migration under fatigue loading, and I believe that MOOSE/MARMOT would be an invaluable tool in achieving this goal.

If it is possible to provide access or guidance on how to obtain the code or permissions for using MOOSE/MARMOT, I would greatly appreciate your assistance. Please let me know if there are any specific procedures or requirements for obtaining access.

Thank you for your time and consideration. I look forward to any guidance or support you can provide, and I am excited about the potential for further advancements in the field of grain boundary migration.

Sincerely,

Wei Peng

**Comment by GiudGiud** [comment]
Hello

Access to all closed source MOOSE-apps developed by INL can be requested here:
https://ncrcaims.inl.gov/Identity/Account/Login

Do you have any collaborators in the US national labs?

While we cannot guarantee a timeline, we are working to open-source MARMOT. Most of MARMOT relies on the MOOSE phase field module which is open-source. 

Guillaume

**Reply by PengWei97** [reply]
Hello Guillaume,

Thank you for your prompt response and for providing the link for requesting access to closed source MARMOT developed by INL.

Regarding your question about collaborators in the US national labs, I must regretfully inform you that I do not currently have any collaborators at US national labs. However, I remain open to potential collaborations in the future, as I believe we could be mutually beneficial.

I also understand that obtaining access to MARMOT may have specific requirements, including affiliation with a US-based institution. I have encountered this limitation while attempting to fill out the access request form. At present, my affiliation is not with a US-based institution, which unfortunately restricts my ability to request access.

Furthermore, I would like to clarify that my current research needs do not require access to the complete MARMOT codebase. Based on the content of the aforementioned paper, I believe that access to the relevant source code mentioned in the paper would suffice for my research purposes. If it is possible for the MOOSE developers to share the specific code related to the concepts discussed in the paper, it would greatly assist me in my research efforts.

I truly appreciate your consideration and assistance in this matter. Access to the specific source code referenced in the paper would be immensely valuable for advancing my research on grain boundary migration and fatigue loading.

Thank you once again for your help, and I look forward to any further guidance or support you can provide.

Best regards,

wei

**Reply by GiudGiud** [reply]
I'll let Marmot developers comment on whether there's anything else we can do outside of the normal processes. 

@laagesen  @dschwen

**Reply by PengWei97** [reply]
Hello @GiudGiud, 

I wanted to follow up on our previous discussion regarding access to the MARMOT source code. In your previous response, you mentioned that you would let the MARMOT developers comment on whether there's anything else that can be done outside of the normal processes.

I would appreciate any further information or updates on this matter. Your assistance in facilitating access to the relevant source code, as mentioned in the paper, would significantly aid my research efforts.

Thank you for your attention to this request, and I look forward to hearing from you or the MARMOT developers.

Best regards,

Wei Peng

**Reply by GiudGiud** [reply]
Hello

Nothing more I can do on my end unfortunately.

Guillaume

**Reply by lindsayad** [reply]
I don’t think there is any capability in marmot that is not in the phase field module. The only thing that might be missing is information related to nuclear materials but you have only spoken thus far about general modeling capability so it seems like you should just move forward why the phase field module.

**Reply by PengWei97** [reply]
Hello @lindsayad ,

In response to your comment, I would like to clarify that the MOOSE phase field module currently lacks specific examples and code implementations of the crystal plasticity-phase field coupling model discussed in Zhao L's paper. The paper clearly states that this coupled model is included in the MOOSE/MARMOT simulation package.

Also, it's worth noting that elastic energy driven grain growth can indeed be found under the path `combined\examples\phase_field-mechanics\`, so I think MARMOT's "Combined" module should contain a similar simulation example. This simulation example uses a crystal plasticity-phase field coupling model.

Thanks for your time and help.

Sincerely,

wei

**Reply by hugary1995** [reply]
@PengWei97 you should contact the corresponding author of that paper and confirm that MARMOT is necessary, i.e. MARMOT has capabilities that MOOSE doesn't have.

**Reply by lindsayad** [reply]
Marmot doesn’t have a combined module

**Reply by lindsayad** [reply]
But it does look like there are some good examples

**Reply by PengWei97** [reply]
Hi @hugary1995, thank you very much for your suggestion. I will send an inquiry email to the corresponding author for verification.

**Reply by PengWei97** [reply]
Hi @lindsayad, thank you for your attention to this matter. If there are any pertinent calculation examples involving the coupling of crystal plasticity and phase field models within MARMOT, I would greatly appreciate it if you could share the corresponding code with me, should conditions allow. This would be immensely beneficial for my ongoing scientific research.

Furthermore, my primary challenge at the moment, as outlined in the conversation below, remains an area where I seek further insights and opinions. Your input is highly valued.

Thank you once again for your assistance.

**Comment by PengWei97** [comment]
Hello @GiudGiud @laagesen @dschwen,

I understand that providing direct access to the MARMOT code may be challenging, considering its current closed-source status. I appreciate the challenges involved in sharing the entire codebase.

As GiudGiud suggested, I'm looking forward to help and answers from MARMOT experts and the community on specific aspects of my research. In the context of a unified description of crystal plasticity and phase field coupled model, specifically discussed in the [zhao_CMS2017](https://www.sciencedirect.com/science/article/pii/S092702561630605X) paper, my main focus is on the calculation of variables and initial state variables  corresponding to activated grains (gr_i) at each integration point.

To elaborate, I perceive two types of non-uniformities in this context: one related to variations between grains, and another associated with variations within individual grains. Handling the former, where different grains interact, appears straightforward by utilizing the weighted averaging method as outlined in the paper.

However, it is the latter aspect, internal variations within a grain, that I seek clarification on. My understanding is that in a crystal plasticity model, at each integration point, there exist unique and well-defined "variables and initial state variables." Yet, the paper mentions the calculation of "initial state variables" for each activated grain gr_i at each integration point, which I find somewhat unclear.

Therefore, I kindly request guidance and insights on addressing this second type of non-uniformity, particularly the calculation of "initial state variables" for each activated grain at individual integration points within a crystal plasticity model. Any assistance, explanation, or references that could shed light on this matter would be greatly appreciated.

I am eager to deepen my understanding of this concept, and your valuable input will contribute significantly to the advancement of my research. Thank you for your time, and I look forward to hearing from the experts and the community on this forum.

Best regards,

Wei peng
